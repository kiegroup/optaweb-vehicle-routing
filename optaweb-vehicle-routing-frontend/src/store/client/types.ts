import { Viewport as LeafletViewport } from 'react-leaflet';
import { Action } from 'redux';

export enum ActionType {
  UPDATE_VIEWPORT = 'UPDATE_VIEWPORT',
  RESET_VIEWPORT = 'RESET_VIEWPORT',
}

export interface UpdateViewportAction extends Action<ActionType.UPDATE_VIEWPORT> {
  value: LeafletViewport;
}

export interface ResetViewportAction extends Action<ActionType.RESET_VIEWPORT> {
}

export interface UserViewport {
  isDirty: boolean;
  center: [number, number];
  zoom: number;
}

export type ViewportAction =
  | UpdateViewportAction
  | ResetViewportAction;
