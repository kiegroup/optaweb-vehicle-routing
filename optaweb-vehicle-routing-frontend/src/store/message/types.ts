import { Action } from 'redux';

export enum ActionType {
  RECEIVE_MESSAGE = 'RECEIVE_MESSAGE',
  READ_MESSAGE = 'READ_MESSAGE',
}

export interface ReceiveMessageAction extends Action<ActionType.RECEIVE_MESSAGE> {
  readonly payload: MessagePayload;
}

export interface ReadMessageAction extends Action<ActionType.READ_MESSAGE> {
  readonly id: string;
}

export type MessageAction = ReceiveMessageAction | ReadMessageAction;

export interface MessagePayload {
  readonly id: string;
  readonly text: string;
}

export interface Message extends MessagePayload {
  readonly status: 'new' | 'read';
}
