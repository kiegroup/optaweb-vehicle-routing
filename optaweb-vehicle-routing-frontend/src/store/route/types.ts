/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Action } from 'redux';

export interface LatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface LatLngWithDescription extends LatLng {
  description: string;
}

export interface Location extends LatLng {
  readonly id: number;
  // TODO decide between optional, nullable and more complex structure (displayName, fullDescription, address, ...)
  readonly description?: string;
}

export interface Vehicle {
  readonly id: number;
  readonly name: string;
  readonly capacity: number;
}

export interface Route {
  readonly vehicle: Vehicle; // TODO change to vehicleId
  readonly visits: Location[];
}

export type LatLngTuple = [number, number];

export interface RouteWithTrack extends Route {
  readonly track: LatLngTuple[];
}

export interface RoutingPlan {
  readonly distance: string;
  readonly vehicles: Vehicle[];
  readonly depot: Location | null;
  readonly visits: Location[];
  readonly routes: RouteWithTrack[];
}

export enum ActionType {
  UPDATE_ROUTING_PLAN = 'UPDATE_ROUTING_PLAN',
  DELETE_LOCATION = 'DELETE_LOCATION',
  ADD_LOCATION = 'ADD_LOCATION',
  ADD_VEHICLE = 'ADD_VEHICLE',
  DELETE_VEHICLE = 'DELETE_VEHICLE',
  CLEAR_SOLUTION = 'CLEAR_SOLUTION',
}

export interface AddLocationAction extends Action<ActionType.ADD_LOCATION> {
  readonly value: LatLngWithDescription;
}

export interface AddVehicleAction extends Action<ActionType.ADD_VEHICLE> {
}

export interface ClearRouteAction extends Action<ActionType.CLEAR_SOLUTION> {
}

export interface DeleteLocationAction extends Action<ActionType.DELETE_LOCATION> {
  readonly value: number;
}

export interface DeleteVehicleAction extends Action<ActionType.DELETE_VEHICLE> {
  readonly value: number;
}

export interface VehicleCapacity {
  vehicleId: number;
  capacity: number;
}

export interface UpdateRouteAction extends Action<ActionType.UPDATE_ROUTING_PLAN> {
  readonly plan: RoutingPlan;
}

export type RouteAction =
  | AddLocationAction
  | AddVehicleAction
  | DeleteLocationAction
  | DeleteVehicleAction
  | UpdateRouteAction
  | ClearRouteAction;
