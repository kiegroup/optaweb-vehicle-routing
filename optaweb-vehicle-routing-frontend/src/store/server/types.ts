import { Action } from 'redux';
import { LatLng } from '../route/types';

export enum ActionType {
  SERVER_INFO = 'SERVER_INFO',
}

export interface ServerInfoAction extends Action<ActionType.SERVER_INFO> {
  value: ServerInfo;
}

export interface Demo {
  name: string;
  visits: number;
}

export type BoundingBox = [LatLng, LatLng];

export interface ServerInfo {
  boundingBox: BoundingBox | null;
  countryCodes: string[];
  demos: Demo[];
}
