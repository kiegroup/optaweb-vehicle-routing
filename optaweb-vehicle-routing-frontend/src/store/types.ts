/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';
import WebSocketClient from 'websocket/WebSocketClient';
import { UserViewport } from './client/types';
import { Demo } from './demo/types';
import { RoutingPlan } from './route/types';
import { ServerInfo } from './server/types';
import { WebSocketConnectionStatus } from './websocket/types';

/**
 * ThunkCommand is a ThunkAction that has no result (it's typically something like
 * `Promise<ActionAfterDataFetched>`, but sending messages over WebSocket usually has no response
 * (with the exception of subscribe), so most of our operations are void).
 *
 * @template A Type of action(s) allowed to be dispatched.
 */
export type ThunkCommand<A extends Action> = ThunkAction<void, AppState, WebSocketClient, A>;

/**
 * Factory method that takes a value and creates an @type {Action}.
 *
 * @template V value type
 * @template A action type
 */
export type ActionFactory<V, A extends Action> = V extends void ?
  // https://stackoverflow.com/questions/55646272/conditional-method-parameters-based-on-generic-type
  () => A : // nullary
  (value: V) => A; // unary

/**
 * Factory method that takes a value and creates a @type {ThunkCommand}.
 *
 * @template V value type
 * @template A action type
 */
export type ThunkCommandFactory<V, A extends Action> = V extends void ?
  () => ThunkCommand<A> : // nullary
  (value: V) => ThunkCommand<A>; // unary

export interface AppState {
  readonly serverInfo: ServerInfo;
  readonly plan: RoutingPlan;
  readonly connectionStatus: WebSocketConnectionStatus;
  readonly demo: Demo;
  readonly userViewport: UserViewport;
}
