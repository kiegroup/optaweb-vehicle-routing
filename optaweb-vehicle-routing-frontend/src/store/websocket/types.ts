import { Action } from 'redux';

export enum WebSocketConnectionStatus {
  OPEN = 'OPEN',
  CLOSED = 'CLOSED',
  ERROR = 'ERROR',
}

export enum ActionType {
  WS_CONNECT = 'WS_CONNECT',
  WS_CONNECT_SUCCESS = 'WS_CONNECT_SUCCESS',
  WS_CONNECT_FAILURE = 'WS_CONNECT_FAILURE',
}

export interface InitWsConnectionAction extends Action<ActionType.WS_CONNECT> {
}

export interface WsConnectionSuccessAction extends Action<ActionType.WS_CONNECT_SUCCESS> {
}

export interface WsConnectionFailureAction extends Action<ActionType.WS_CONNECT_FAILURE> {
  readonly error: string;
}

export type WebSocketAction =
  | InitWsConnectionAction
  | WsConnectionFailureAction
  | WsConnectionSuccessAction;
