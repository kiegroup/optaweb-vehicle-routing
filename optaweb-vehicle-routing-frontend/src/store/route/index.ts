import * as routeOperations from './operations';
import { routeReducer } from './reducers';
import * as routeSelectors from './selectors';

export { routeOperations, routeSelectors };

export default routeReducer;
