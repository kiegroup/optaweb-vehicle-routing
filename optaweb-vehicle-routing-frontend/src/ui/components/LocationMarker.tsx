/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as L from 'leaflet';
import * as React from 'react';
import { Marker, Tooltip } from 'react-leaflet';
import { Location } from 'store/route/types';

const homeIcon = L.icon({
  iconAnchor: [12, 12],
  iconSize: [24, 24],
  iconUrl: 'if_big_house-home_2222740.png',
  popupAnchor: [0, -10],
  shadowAnchor: [16, 2],
  shadowSize: [50, 16],
  shadowUrl: 'if_big_house-home_2222740_shadow.png',
});

const defaultIcon = new L.Icon.Default();

export interface Props {
  location: Location;
  isDepot: boolean;
  isSelected: boolean;
  removeHandler: (id: number) => void;
}

const LocationMarker: React.FC<Props> = ({
  location,
  isDepot,
  isSelected,
  removeHandler,
}) => {
  const icon = isDepot ? homeIcon : defaultIcon;
  return (
    <Marker
      key={location.id}
      position={location}
      icon={icon}
      onClick={() => removeHandler(location.id)}
    >
      <Tooltip
        // `permanent` is a static property (this is a React-Leaflet-specific
        // approach: https://react-leaflet.js.org/docs/en/components). Changing `permanent` prop
        // doesn't result in calling `setPermanent()` on the Leaflet element after the Tooltip component is mounted.
        // We're using `key` to force re-rendering of Tooltip when `isSelected` changes. A similar use case for
        // the `key` property is described here:
        // https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html
        // #recommendation-fully-uncontrolled-component-with-a-key
        key={isSelected ? 'selected' : ''}
        permanent={isSelected}
      >
        {`Location ${location.id} [Lat=${location.lat}, Lng=${location.lng}]`}
      </Tooltip>
    </Marker>
  );
};

export default LocationMarker;
