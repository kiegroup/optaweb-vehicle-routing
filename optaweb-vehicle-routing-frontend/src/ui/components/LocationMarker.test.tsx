/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Marker } from 'react-leaflet';
import { Location } from 'store/route/types';
import LocationMarker, { Props } from './LocationMarker';

const location: Location = {
  id: 1,
  lat: 1.345678,
  lng: 1.345678,
};

describe('Location Marker', () => {
  it('render depot', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      isDepot: true,
      isSelected: false,
      location,
    };
    const locationMarker = shallow(<LocationMarker {...props} />);
    expect(toJson(locationMarker)).toMatchSnapshot();
  });

  it('render visit', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      isDepot: false,
      isSelected: false,
      location,
    };
    const locationMarker = shallow(<LocationMarker {...props} />);
    expect(toJson(locationMarker)).toMatchSnapshot();
  });

  it('selected visit should show a tooltip', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      isDepot: false,
      isSelected: true,
      location,
    };
    const locationMarker = shallow(<LocationMarker {...props} />);
    expect(toJson(locationMarker)).toMatchSnapshot();
  });

  it('should call remove handler when clicked', () => {
    const props: Props = {
      removeHandler: jest.fn(),
      isDepot: false,
      isSelected: true,
      location,
    };
    const locationMarker = shallow(<LocationMarker {...props} />);
    locationMarker.find(Marker).simulate('click');
    expect(props.removeHandler).toBeCalled();
  });
});
