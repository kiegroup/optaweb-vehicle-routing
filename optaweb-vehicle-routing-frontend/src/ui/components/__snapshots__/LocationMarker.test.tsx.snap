// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location Marker render depot 1`] = `
<ForwardRef(Leaflet(Marker))
  icon={
    NewClass {
      "_initHooksCalled": true,
      "options": Object {
        "iconAnchor": Array [
          12,
          12,
        ],
        "iconSize": Array [
          24,
          24,
        ],
        "iconUrl": "if_big_house-home_2222740.png",
        "popupAnchor": Array [
          0,
          -10,
        ],
        "shadowAnchor": Array [
          16,
          2,
        ],
        "shadowSize": Array [
          50,
          16,
        ],
        "shadowUrl": "if_big_house-home_2222740_shadow.png",
      },
    }
  }
  key="1"
  onClick={[Function]}
  position={
    Object {
      "id": 1,
      "lat": 1.345678,
      "lng": 1.345678,
    }
  }
>
  <ForwardRef(Leaflet(Tooltip))
    key=""
    permanent={false}
  >
    Location 1 [Lat=1.345678, Lng=1.345678]
  </ForwardRef(Leaflet(Tooltip))>
</ForwardRef(Leaflet(Marker))>
`;

exports[`Location Marker render visit 1`] = `
<ForwardRef(Leaflet(Marker))
  icon={
    NewClass {
      "_initHooksCalled": true,
      "options": Object {},
    }
  }
  key="1"
  onClick={[Function]}
  position={
    Object {
      "id": 1,
      "lat": 1.345678,
      "lng": 1.345678,
    }
  }
>
  <ForwardRef(Leaflet(Tooltip))
    key=""
    permanent={false}
  >
    Location 1 [Lat=1.345678, Lng=1.345678]
  </ForwardRef(Leaflet(Tooltip))>
</ForwardRef(Leaflet(Marker))>
`;

exports[`Location Marker selected visit should show a tooltip 1`] = `
<ForwardRef(Leaflet(Marker))
  icon={
    NewClass {
      "_initHooksCalled": true,
      "options": Object {},
    }
  }
  key="1"
  onClick={[Function]}
  position={
    Object {
      "id": 1,
      "lat": 1.345678,
      "lng": 1.345678,
    }
  }
>
  <ForwardRef(Leaflet(Tooltip))
    key="selected"
    permanent={true}
  >
    Location 1 [Lat=1.345678, Lng=1.345678]
  </ForwardRef(Leaflet(Tooltip))>
</ForwardRef(Leaflet(Marker))>
`;
